src/Core/Grid/Query/CartQueryBuilder.php

```diff
@@ -0,0 +1,309 @@
+    private function getQueryBuilder(array $filters)
+    {
+        $qbOnline = $this->connection
+            ->createQueryBuilder()
+            ->select('DISTINCT co.`id_guest`')
+            ->from($this->dbPrefix . 'connections', 'co')
+            ->where('TIME_TO_SEC(TIMEDIFF(\'' . pSQL(date('Y-m-d H:i:00', time())) . '\', `date_add`)) < ' . self::CUSTOMER_ONLINE_TIME);
+
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->from($this->dbPrefix . 'cart', 'c');
+
+        $qb->leftJoin(
+            'c',
+            $this->dbPrefix . 'carrier',
+            'ca',
+            'c.`id_carrier` = ca.`id_carrier`'
+        );
+
+        $qb->leftJoin(
+            'c',
+            $this->dbPrefix . 'orders',
+            'o',
+            'c.`id_cart` = o.`id_cart`'
+        );
+
+        $qb->leftJoin(
+            'c',
+            $this->dbPrefix . 'customer',
+            'cu',
+            'c.`id_customer` = cu.`id_customer`'
+        );
+
+        $qb->leftJoin(
+            'c',
+            '(' . $qbOnline->getSQL() . ')',
+            'co',
+            'co.`id_guest` = c.`id_guest`'
+        );
+
+        $qb->leftJoin(
+            'c',
+            $this->dbPrefix . 'shop',
+            's',
+            'c.`id_shop` = s.`id_shop`'
+        );
+
+        if ($this->multistoreContextChecker->isSingleShopContext()) {
+            $qb->andWhere('s.`id_shop` = :shopId');
+            $qb->setParameter('shopId', $this->contextShopId);
+        } elseif ($this->multistoreContextChecker->isGroupShopContext()) {
+            $qb->andWhere('s.`id_shop_group` = :shopGroupId');
+            $qb->setParameter('shopGroupId', $this->contextShopGroupId);
+        }
+
+        $this->applyFilters($qb, $filters);
+
+        return $qb;
+    }
+
+    /**
+     * Apply filters to carts query builder
+     *
+     * @param QueryBuilder $qb
+     * @param array $filters
+     *
+     * @return void
+     */
+    private function applyFilters(QueryBuilder $qb, array $filters): void
+    {
+        $allowedFilters = [
+            'id_cart',
+            'id_order',
+            'customer_name',
+            'carrier_name',
+            'date_add',
+            'customer_online',
+        ];
+
+        foreach ($filters as $filterName => $filterValue) {
+            if (!in_array($filterName, $allowedFilters)) {
+                continue;
+            }
+
+            if ('id_cart' === $filterName) {
+                $qb->andWhere('c.id_cart = :' . $filterName);
+                $qb->setParameter($filterName, $filterValue);
+                continue;
+            }
+
+            if ('id_order' === $filterName) {
+                $filterValue = strtolower($filterValue);
+                $strNonOrdered = strtolower($this->translator->trans('Non ordered', [], 'Admin.Orderscustomers.Feature'));
+                $strAbandonedCart = strtolower($this->translator->trans('Abandoned cart', [], 'Admin.Orderscustomers.Feature'));
```

Comment:
I dont really understand what happens here, why you need these translations? how id_order value becomes string? (it should be number)

Reply:
In fact, it's aim to reproduce legacy behavior but I'm not a big fan too!
In legacy page, we have only one column for id_order and if cart haven't id_order we display "Not ordered" or "Abandoned cart". We have also a input field to search on this column, and for instance If we type "Not" or "Not ordered" we must display all cart "Not ordered".
Personally, I prefer to add another column for cart status and so provide checkbox list and so we can remove translators things in this class! What do you thing ? :)

Reply:
My solution are implemented now in my PR :)